### PR TITLE
Add FileSystemGet Command

### PR DESCRIPTION
**Pull request by rjesquivias**
Introduced the Communicator & implementation/test code for FileSystemGet

### CODE REVIEW

File: src/FileSystemReceiver/FileSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        if(returnCode != 0) return false;
+    }
+    
+    return true;
+}
+
+int FileSystem::readIntoBuffer(char* buffer, int bufSize, int readPosition, std::string fileName)
+{
+    std::ifstream is(fileName, std::ifstream::binary);
+    if(!is.is_open()) return -1; // TODO: Throw Error
```
Comment (crgodfirnon) [inline src/FileSystemReceiver/FileSystem.cpp]: ugh

File: src/Communicator/NetworkCommunicator.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include "ICommunicator.h"
+
+class NetworkCommunicator : public ICommunicator
+{
+public:
+    NetworkCommunicator() {} 
+
+    virtual ReturnTypes send(char* data, int size){ return ReturnTypes::SUCCESS; }
```
Comment (crgodfirnon) [inline src/Communicator/NetworkCommunicator.h]: override

File: src/FileSystemReceiver/WindowsFileSystemReceiver.cpp [hunk 3]
```diff
@@ -86,4 +86,30 @@ ReturnTypes WindowsFileSystemReceiver::execute(std::string file)
...
     bool status = fileSystem->isAbsolutePath(file) ? fileSystem->execute(file) : fileSystem->execute(getDirectory() + file);
     return status ? ReturnTypes::SUCCESS : ReturnTypes::FAILURE;
+}
+
+ReturnTypes WindowsFileSystemReceiver::getFile(std::string fileName) const
+{
+    if(!fileSystem->isAbsolutePath(fileName))
+        fileName = getDirectory() + fileName;
+
+    int bufSize = 1024; // TODO: Set to variable
```
Comment (crgodfirnon) [inline src/FileSystemReceiver/WindowsFileSystemReceiver.cpp]: 1024 should be a constant somewhere

File: src/FileSystemReceiver/WindowsFileSystemReceiver.h [hunk 4]
```diff
@@ -6,14 +6,15 @@
 class WindowsFileSystemReceiver : public FileSystemReceiver
 {
 public:
-    WindowsFileSystemReceiver(std::shared_ptr<IFileSystem> fileSystem) { this->fileSystem = fileSystem; }
+    WindowsFileSystemReceiver(std::shared_ptr<IFileSystem> fileSystem, std::shared_ptr<ICommunicator> communication) 
+                             { this->fileSystem = fileSystem; this->communication = communication; }
```
Comment (crgodfirnon) [inline src/FileSystemReceiver/WindowsFileSystemReceiver.h]: : fileSystem(fileSystem), communication(communication) 
{
}

File: src/FileSystemReceiver/WindowsFileSystemReceiver.h [hunk 5]
```diff
@@ -6,14 +6,15 @@
...
     ~WindowsFileSystemReceiver() { }
 
-	ReturnTypes listDirectory() { return ReturnTypes::SUCCESS;  }
+	ReturnTypes listDirectory() { return ReturnTypes::SUCCESS; }
```
Comment (crgodfirnon) [inline src/FileSystemReceiver/WindowsFileSystemReceiver.h]: ur spacing is fucked here

File: src/main.cpp [hunk 6]
```diff
@@ -1,12 +1,15 @@
...
+#include "Communicator/NetworkCommunicator.h"
 using namespace std;
 
-
 int main()
 {
-    WindowsFileSystemReceiver fsr(nullptr);
-    
+    std::shared_ptr<IFileSystem> fileSystem = std::make_shared<FileSystem>();
+	std::shared_ptr<ICommunicator> communicator = std::make_shared<NetworkCommunicator>();
```
Comment (crgodfirnon) [inline src/main.cpp]: spacing

File: test/src/Communicator/MockCommunicator.cpp [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#include "MockCommunicator.h"
+
+ReturnTypes MockCommunicator::send(char* data, int size)
+{
+    for(int i = 0; i < size; i++)
+    {
+        if(historyIndex >= 1232) historyIndex = 0; // Loop back around and start rerecording data
```
Comment (crgodfirnon) [inline test/src/Communicator/MockCommunicator.cpp]: can you make this 1232 a constant somewhere?

File: test/src/FileSystemReceiverTest/FileSystemReceiverTest.cpp [hunk 8]
```diff
@@ -1,35 +1,15 @@
...
-
-    fsr.listDirectory();
-}
-
 // Windows Specific Tests
 TEST(WindowsFileSystemReceiverTest, TestChangeDirectoryCommand)
 {
     std::shared_ptr<IFileSystem> fileSystem = std::make_shared<MockFileSystem>();
-    WindowsFileSystemReceiver fsr(fileSystem);
+	std::shared_ptr<ICommunicator> communicator = std::make_shared<MockCommunicator>();
```
Comment (crgodfirnon) [inline test/src/FileSystemReceiverTest/FileSystemReceiverTest.cpp]: spacing

File: test/src/FileSystemReceiverTest/FileSystemReceiverTest.cpp [hunk 9]
```diff
@@ -86,12 +66,17 @@ TEST(WindowsFileSystemReceiverTest, TestChangeDirectoryCommand)
...
+	//fsr.changeDirectory("C:\\Users\\user\\Desktop");
+	//ASSERT_EQ(fsr.getDirectory(), "C:\\Users\\user\\Desktop\\");
 }
 
 TEST(WindowsFileSystemReceiverTest, TestExecuteCommand)
 {
     std::shared_ptr<IFileSystem> fileSystem = std::make_shared<MockFileSystem>();
-    WindowsFileSystemReceiver fsr(fileSystem);
+	std::shared_ptr<ICommunicator> communicator = std::make_shared<MockCommunicator>();
+    WindowsFileSystemReceiver fsr(fileSystem, communicator);
```
Comment (crgodfirnon) [inline test/src/FileSystemReceiverTest/FileSystemReceiverTest.cpp]: spacing

File: test/src/FileSystemReceiverTest/MockFileSystem.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        int integerValue = (readPosition + i) / 4 ; // read into buffer 8 bits at a time
+
+        buffer[i++] = (integerValue & 0xFF);
+        buffer[i++] = ((integerValue >> 8) & 0xFF);
+        buffer[i++] = ((integerValue >> 16) & 0xFF);
+        buffer[i] = ((integerValue >> 24) & 0xFF); // Let the loop do the last increment
+    }
+
```
Comment (crgodfirnon) [inline test/src/FileSystemReceiverTest/MockFileSystem.cpp]: can any of these be constants?

File: src/FileSystemReceiver/FileSystem.cpp (hunk 1 above)
Comment (rjesquivias) [reply src/FileSystemReceiver/FileSystem.cpp]: I mean I'm not writing that shit now

File: src/Communicator/NetworkCommunicator.h (hunk 2 above)
Comment (rjesquivias) [reply src/Communicator/NetworkCommunicator.h]: done

File: src/FileSystemReceiver/WindowsFileSystemReceiver.cpp (hunk 3 above)
Comment (rjesquivias) [reply src/FileSystemReceiver/WindowsFileSystemReceiver.cpp]: Added into FileSystemGet.h

File: test/src/Communicator/MockCommunicator.cpp (hunk 7 above)
Comment (rjesquivias) [reply test/src/Communicator/MockCommunicator.cpp]: Done, this is still kinda gross cause the values in MockCommunicator.h & MockFileSystem.cpp need to match and I don't want to pass this value between them, might look into something better later.

File: test/src/FileSystemReceiverTest/MockFileSystem.cpp (hunk 10 above)
Comment (rjesquivias) [reply test/src/FileSystemReceiverTest/MockFileSystem.cpp]: Created a mask

**Pull request merged**